### PR TITLE
[IMP] mrp,stock: mrp back to basics 9

### PR DESCRIPTION
**Pull request by R1D1CUL0US**
Combination of multiple small improvements in MRP, see different feature commits for details.

Task: 2479234

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/stock_move.py [hunk 1]
```diff
@@ -398,7 +398,7 @@ def action_show_details(self):
     def _action_cancel(self):
         res = super(StockMove, self)._action_cancel()
         mo_to_cancel = self.mapped('raw_material_production_id').filtered(lambda p: all(m.state == 'cancel' for m in p.move_raw_ids))
-        if mo_to_cancel:
+        if mo_to_cancel and not self.env.context.get('skip_consumption'):
```
Comment (amoyaux) [inline addons/mrp/models/stock_move.py:401]: You could explain why you need to check this context key in the message commit

File: addons/analytic/views/analytic_line_views.xml [hunk 2]
```diff
@@ -8,9 +8,9 @@
                 <field name="company_id" invisible="1"/>
                 <field name="product_uom_category_id" invisible="1"/>
                 <field name="date" optional="show"/>
-                <field name="name"/>
-                <field name="account_id"/>
-                <field name="plan_id"/>
+                <field name="name" optional="show"/>
+                <field name="account_id" optional="show"/>
+                <field name="plan_id" optional="show"/>
```
Comment (h4818) [inline addons/analytic/views/analytic_line_views.xml:13]: OK for me if @tsb-odoo or @gawa-odoo is ok.
Seems weird to do this on the default view, should we not have a separate view for MRP analytic margin report?
Many users might activate these if they use analytic accounts for anything else.

File: addons/stock/models/stock_scrap.py [hunk 3]
```diff
@@ -139,8 +139,8 @@ def do_scrap(self):
             scrap.name = self.env['ir.sequence'].next_by_code('stock.scrap') or _('New')
             move = self.env['stock.move'].create(scrap._prepare_move_values())
             # master: replace context by cancel_backorder
-            move.with_context(is_scrap=True)._action_done()
-            scrap.write({'move_id': move.id, 'state': 'done'})
+            moves = move.with_context(is_scrap=True)._action_done()
+            scrap.write({'move_ids': [move.id for move in moves], 'state': 'done'})
```
Comment (amoyaux) [inline addons/stock/models/stock_scrap.py:144]: you could set scrap_id in _prepare_move_values. Instead of writing the one2many field it will prevent to trigger 2 write instead if 1

File: addons/stock/models/stock_scrap.py [hunk 4]
```diff
@@ -151,16 +151,12 @@ def action_get_stock_picking(self):
...
-        self.ensure_one()
-        if float_is_zero(self.scrap_qty,
-                         precision_rounding=self.product_uom_id.rounding):
-            raise UserError(_('You can only enter positive quantities.'))
+    def check_available_qty(self):
         if self.product_id.type != 'product':
-            return self.do_scrap()
+            return True
         precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
         available_qty = sum(self.env['stock.quant']._gather(self.product_id,
```
Comment (amoyaux) [inline addons/stock/models/stock_scrap.py:177]: Maybe you could try with using the qty_available product product. It would avoid the override in mrp

File: addons/mrp/models/stock_scrap.py [hunk 5]
```diff
@@ -50,3 +57,37 @@ def _onchange_serial_number(self):
...
                 return super()._onchange_serial_number()
+
+    @api.depends('move_ids', 'move_ids.move_line_ids.qty_done', 'product_id')
+    def _compute_scrap_qty(self):
+        self.scrap_qty = 1
+        for scrap in self:
+            if not scrap.bom_id:
+                return super(StockScrap, scrap)._compute_scrap_qty()
+            if scrap.move_ids:
+                minimum = 0
```
Comment (amoyaux) [inline addons/mrp/models/stock_scrap.py:68]: Isn't it possible to use _compute_kit_quantities from stock.move?

File: addons/mrp/models/stock_scrap.py [hunk 6]
```diff
@@ -50,3 +57,37 @@ def _onchange_serial_number(self):
...
+            available_qty = sum(self.env['stock.quant']._gather(bom_line.product_id,
+                                                                self.location_id,
+                                                                self.lot_id,
+                                                                self.package_id,
+                                                                self.owner_id,
+                                                                strict=True).mapped('quantity'))
+            scrap_qty = bom_line.product_uom_id._compute_quantity(self.scrap_qty * bom_line.product_qty, bom_line.product_id.uom_id)
+            if not float_compare(available_qty, scrap_qty, precision_digits=precision) >= 0:
+                return False
+        return True
```
[the remark this answers is not in the document]
Comment (R1D1CUL0US) [reply addons/mrp/models/stock_scrap.py]: You can't completely avoid an override (I think) but I have refactored to use available_qty and only override the new `_should_check_available_qty` in mrp/stock_scrap

File: addons/analytic/views/analytic_line_views.xml (hunk 2 above)
Comment (gawa-odoo) [reply addons/analytic/views/analytic_line_views.xml:13]: Indeed, it seems weird to do so for all analytic lines. 
If someone hides it for the view coming from 'Gross margin', it does not mean he wants to hide for the general view of analytic lines. (And I find especially weird if you hide the analytic account for the general view where we group by analytic account by default)

File: addons/mrp/models/mrp_workorder.py [hunk 7]
```diff
@@ -115,9 +115,9 @@ def _read_group_workcenter_id(self, workcenters, domain, order):
         'stock.move.line', 'workorder_id', 'Moves to Track',
         help="Inventory moves for which you must scan a lot number at this work order")
     finished_lot_id = fields.Many2one(
-        'stock.lot', string='Lot/Serial Number', compute='_compute_finished_lot_id',
-        inverse='_set_finished_lot_id', domain="[('product_id', '=', product_id), ('company_id', '=', company_id)]",
-        check_company=True)
+        'stock.lot', string='Lot/Serial Number', related='production_id.lot_producing_id',
```
Comment (amoyaux) [inline addons/mrp/models/mrp_workorder.py:118]: The line in the commit message "Also add the production end date to form and list view." seems not related to this commit

File: addons/mrp/tests/test_order.py [hunk 8]
```diff
@@ -1322,45 +1322,6 @@ def test_product_produce_12(self):
         mo.move_finished_ids._action_done()
         mo.button_mark_done()
 
-    def test_product_produce_13(self):
```
Comment (amoyaux) [inline addons/mrp/tests/test_order.py:1325]: It would be great to create a new test for the new behavior

File: addons/mrp/models/stock_scrap.py [hunk 9]
```diff
@@ -14,6 +13,12 @@ class StockScrap(models.Model):
         'mrp.workorder', 'Work Order',
         states={'done': [('readonly', True)]},
         check_company=True) # Not to restrict or prefer quants, but informative
+    product_is_kit = fields.Boolean(related='product_id.is_kits')
+    product_template = fields.Many2one(related='product_id.product_tmpl_id')
+    bom_id = fields.Many2one(
+        'mrp.bom', 'Kit',
+        domain="[('type', '=', 'phantom'), ('product_tmpl_id', '=', product_template)]",
```
Comment (amoyaux) [inline addons/mrp/models/stock_scrap.py:20]: I would add a condition to ensure the BoM doesn't have a product define or the product define is the same than the scrap product

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
robodoo rebase-ff

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Comment by amoyaux** [comment]
robodoo r- just waiting for the cleaning of precompute on sml

**Comment by ticodoo** [comment]
@robodoo r+
